### PR TITLE
Support matrix population mixing in convolution scanner

### PR DESCRIPTION
**Pull request by damonbayer**
## Summary

- support matrix-valued multipliers in `new_convolve_scanner`
- document `M[i, j]` generically as the contribution of component `j` to component `i`, while preserving scalar and vector behavior
- add regression tests for matrix multiplication and off-diagonal transmission
- add an executable Quarto tutorial that explicitly presents population transmission as one application of the generic utility, covering the coupled renewal equation, `M[target, source]`, constant and time-varying mixing, and two-population figures
- link the tutorial from the documentation sidebar and landing page

## Testing

- `uv run pytest -m 'not integration'` (843 passed, 35 deselected)
- `uv run pre-commit run --files docs/index.md docs/tutorials/.pages docs/tutorials/population_mixing.qmd pyrenew/convolve.py pyrenew/latent/infection_functions.py test/test_convolve.py`
- `QUARTO_PYTHON=.venv/bin/python quarto render docs/tutorials/population_mixing.qmd`

### CODE REVIEW

File: pyrenew/convolve.py [hunk 1]
```diff
@@ -47,42 +47,55 @@ def new_convolve_scanner(
         [`jax.lax.scan`][] or
         [`numpyro.contrib.control_flow.scan`][]
         for convolution.
-        This function takes a history subset array and
-        a scalar, computes the dot product of
-        the supplied convolution array with the history
-        subset array, multiplies by the scalar, and
+        This function takes a history subset array and a scalar,
+        vector, or matrix, computes the dot product of the supplied
```
Comment (dylanhmorris) [inline pyrenew/convolve.py]:
```suggestion
        This function takes a history subset array and a multiplier (which can be a scalar,
        a vector, or a matrix). It computes the dot product of the supplied
```

File: pyrenew/convolve.py [hunk 2]
```diff
@@ -47,42 +47,55 @@ def new_convolve_scanner(
...
         array formed by the 2nd-through-last entries
         of the old history subset array followed by that same
         resulting value.
 
     Notes
     -----
-    The following iterative operation is found often
-    in renewal processes:
+    The following vector-valued iterative operation generalizes the
+    scalar operation previously supported by this function:
```
Comment (dylanhmorris) [inline pyrenew/convolve.py]:
```suggestion
    The following iterative operation is common in renewal processes:
```

File: pyrenew/convolve.py [hunk 3]
```diff
@@ -47,42 +47,55 @@ def new_convolve_scanner(
...
-    $m(t)$ is a scalar for each value of time $t$,
-    and $f$ is a scalar-valued function.
-
-    Given $\mathbf{d}$, and optionally $f$,
-    this factory function returns a new function that
-    performs one step of this process while scanning along
-    an array of  multipliers (i.e. an array
-    giving the values of $m(t)$) using [`jax.lax.scan`][].
+    Here $\mathbf{d}$ is a vector of length $n$, $\mathbf{X}(t)$
+    contains $K$ components, $M(t)$ is a $K \times K$ matrix, and $f$
```
Comment (dylanhmorris) [inline pyrenew/convolve.py]:
```suggestion
    Here $\mathbf{d}$ is a vector of length $n$, each $\mathbf{X}(t)$
    is a vector of length $k$, $M(t)$ is a $k \times k$ matrix, and $f$
```

File: pyrenew/convolve.py [hunk 4]
```diff
@@ -47,42 +47,55 @@ def new_convolve_scanner(
...
+    acts on the resulting vector. The entry $M_{ij}(t)$ determines the
+    contribution of component $j$ to the update of component $i$.
+    This function assigns no particular interpretation to the components
+    or multipliers; those depend on the calling model.
```
Comment (dylanhmorris) [inline pyrenew/convolve.py]:
```suggestion
    acts on the resulting vector.
```

File: pyrenew/convolve.py [hunk 5]
```diff
@@ -47,42 +47,55 @@ def new_convolve_scanner(
...
+
+    Given $\mathbf{d}$, and optionally $f$, this factory function
+    returns a new function that performs one step of this process while
+    scanning along an array of matrices giving the values of $M(t)$
+    using [`jax.lax.scan`][]. Scalar and vector multipliers remain
+    supported; these act elementwise on the convolved history.
```
Comment (dylanhmorris) [inline pyrenew/convolve.py]:
```suggestion
    using [`jax.lax.scan`][]. 
    
    When scanning along an array of scalar multipliers $m(t)$ or vector
    multipliers $\mathbf{m}(t)$, the function performs performs elementwise 
    multiplication by $m$. This is equivalent to treating a scalar $m(t)$ or vector 
    $\mathbf{m}(t)$ as implying a diagonal matrix multiplier $M(t)$ whose 
    diagonal values  are all equal to $m(t)$ or are given by $\mathbf{m}(t)$, 
    respectively.
```

File: pyrenew/latent/infection_functions.py [hunk 6]
```diff
@@ -27,7 +27,10 @@ def compute_infections_from_rt(
         same length as the generation interval
         pmf vector.
     Rt
-        Timeseries of $\\mathcal{R}(t)$ values
+        Timeseries of $\\mathcal{R}(t)$ values. Each timepoint can be a
+        scalar shared by all populations, a vector of population-specific
+        values, or a matrix whose entry ``[i, j]`` represents transmission
+        from source population ``j`` to target population ``i``.
```
Comment (dylanhmorris) [inline pyrenew/latent/infection_functions.py]:
```suggestion
        values, or a matrix whose entry ``[i, j]`` represents $\\mathcal{R}_{j \to i}(t)$:
        the expected number of new infections in population $i$ caused by an infectious 
        individual in population $j$ at time $t$.
```

**Review by dylanhmorris — COMMENTED** [review]
Various wording tweaks

**Comment by cdc-mitzimorris** [comment]
The matrix-convolution support in this PR is a necessary building block for contact mixing in PyRenew. Before presenting contact mixing as a supported feature, however, we should discuss how it fits into the high-level API.

I may be missing context from while I was away. Could we clarify:

- Is the goal a low-level primitive or a fully supported contact-mixing feature?
- Is the matrix the complete next-generation operator, or should it compose with the existing subpopulation Rt process?
- Is mixing fixed, inferred, or supplied per dataset?
- Should mixing allow asymmetric directionality, and are any reciprocity constraints required?

These choices affect both the epidemiological interpretation and the public API. Even when the modeling need is urgent, a short issue or design note would help us record the use case, agree on these decisions, and keep the implementation consistent with the rest of PyRenew.

I suggest merging the low-level primitive once its semantics and validation are settled, deferring or relabeling the tutorial, and opening a short design issue for integration with SubpopulationInfections, PyrenewBuilder, and MultiSignalModel. We can add the end-user tutorial once that path exists.

I’m happy to draft the design issue and work through the interface together.

File: docs/tutorials/.pages [hunk 7]
```diff
@@ -3,6 +3,7 @@ nav:
   - building_multisignal_models.md
   - latent_infections.md
   - latent_subpopulation_infections.md
+  - population_mixing.md
```
Comment (cdc-mitzimorris) [inline docs/tutorials/.pages:6]: Defer this tutorial until we have a high-level API.

File: pyrenew/convolve.py [hunk 8]
```diff
@@ -38,51 +38,78 @@ def new_convolve_scanner(
...
-        history_subset: ArrayLike, multiplier: float
-    ) -> tuple[ArrayLike, float]:  # numpydoc ignore=GL08
-        new_val = transform(
-            multiplier * jnp.einsum("i...,i...->...", array_to_convolve, history_subset)
+        history_subset: ArrayLike, multiplier: ArrayLike
+    ) -> tuple[ArrayLike, ArrayLike]:  # numpydoc ignore=GL08
+        convolved_history = jnp.einsum(
+            "i...,i...->...", array_to_convolve, history_subset
         )
+        if jnp.ndim(multiplier) < 2:
```
Comment (cdc-mitzimorris) [inline pyrenew/convolve.py:106]: Could we make matrix application explicit rather than infer it solely from ndim? Previously, higher-dimensional multipliers were applied elementwise; this now interprets every rank-2-or-higher multiplier as a matrix. That can break callers whose state has multiple batch dimensions. A separate matrix scanner or an explicit mode would preserve the existing behavior and make the contract unambiguous.

File: pyrenew/latent/infection_functions.py [hunk 9]
```diff
@@ -27,7 +27,11 @@ def compute_infections_from_rt(
         same length as the generation interval
         pmf vector.
     Rt
-        Timeseries of $\\mathcal{R}(t)$ values
+        Timeseries of $\\mathcal{R}(t)$ values. Each timepoint can be a
```
Comment (cdc-mitzimorris) [inline pyrenew/latent/infection_functions.py:30]: Could we describe this as an effective reproduction or next-generation matrix and distinguish it from a raw contact matrix? As used here, each entry already represents transmission from a source to a target. Survey contact matrices generally require scaling for transmissibility, susceptibility/infectiousness, population sizes, and sometimes reciprocity before they can be used in this equation. Clarifying that distinction would help prevent users from passing raw contact data directly as Rt.

File: test/test_convolve.py [hunk 10]
```diff
@@ -157,26 +157,129 @@ def test_double_convolve_scanner_using_scan(arr1, arr2, history, m1, m2, transfo
...
+    scanner = pc.new_convolve_scanner(arr, transform)
+    mult_mat = multiplier * jnp.eye(history.shape[1])
+    assert jnp.ndim(mult_mat) > 1
+    latest_mat, new_val_mat = scanner(history, mult_mat)
+    latest_vec, new_val_vec = scanner(history, multiplier)
+    assert_array_equal(latest_mat, latest_vec)
+    assert_array_equal(new_val_mat, new_val_vec)
+
+
+def test_convolve_scanner_with_population_mixing_matrix():
```
Comment (cdc-mitzimorris) [inline test/test_convolve.py:250]: Could we test matrix-valued input through compute_infections_from_rt, rather than only through the scanner? I would also like coverage for a multi-step time-varying matrix against an independent calculation, invalid matrix/state shapes, and JIT/gradient compatibility.

**Review by cdc-mitzimorris — COMMENTED** [review]
These three comments about the multiplier dispatch, the contact-versus-reproduction-matrix interpretation, and coverage of the public function are separate from the high-level API questions in my earlier comment.

File: pyrenew/convolve.py (hunk 8 above)
Comment (dylanhmorris) [reply pyrenew/convolve.py:106]: This is a good argument for moving away from ad-hoc scanner factors to something more structured (#177). I have sketched out a potential approach in this draft PR:
https://github.com/CDCgov/PyRenew/pull/878
